### ISSUE TITLE
Tablets serving stale data after backup

### ISSUE THREAD

**Issue by setassociative** [vitessio/vitess#4426]
#### Overview of the Issue

(credit to @zmagg for the initial writeup that is ~reproduced here)

When a tablet completes creating a backup it returns to a `SERVING` state without consideration of accrued replication lag. If this value larger than you're comfortable serving in then there is a period of time between completing the backup and detecting a tablet in this unhealthy state where you're serving stale data.

The fallout is unnecessarily noisy alerts and potentially serving bad data.

#### Reproduction Steps

Roughly:

1. Begin backup that takes a non trivial period of time
2. Master receives enough writes that replication is slow
3. Observe tablet serving queries during replication while lag > an unhealthy threshold
4. Healthchecks notice unhealthy state and tablet stops serving queries

#### Log Fragments

```
15:00:54.948165 tabletserver.go:312] TabletServer state: NOT_SERVING (Not Connected) -> NOT_SERVING (Transitioning)
15:00:54.970844 engine.go:183] Time taken to load the schema: 1.053µs
15:00:54.973905 tabletserver.go:312] TabletServer state: NOT_SERVING (Transitioning) -> SERVING
15:00:54.973935 updatestreamctl.go:213] Enabling update stream, dbname: vt_byteam
15:00:54.973985 state_change.go:154] Done with post-action state refresh
15:00:54.975066 rpc_server.go:93] TabletManager.Backup(concurrency:4 )(on us_east_1-0100125104 from ): <nil>
15:00:55.977517 healthcheck.go:269] Disabling query service because of health-check failure: reported replication lag: 3049 higher than unhealthy threshold: 300
15:00:55.977543 tabletserver.go:312] TabletServer state: SERVING -> SHUTTING_DOWN (Shutting Down)
```

#### Suggested Fix

Tablets will transition into `NOT_SERVING` once completing a backup and we'll let health checks determine that it's safe to move them into a serving state when lag < healthy threshold.

**Comment by setassociative** [comment]
**tl;dr:** I took a look at the suggested fix which attempts to sidestep the stale start up by explicitly setting serving state to false before switching back to the original (serving) tablet type. That didn't work so I did some digging and hypothesizing. Caveats should be applied that I'm not super familiar with this codebase.

### Initial Goal: Test the suggested fix

In order to replicate the bad state I passed `degraded_threshold="7s"` & `unhealthy_threshold="15s"` to the `vttablet` process. This allowed me to hit the desired state by just initiating a tablet backup as the process on my test data took about a minute during which I accumulated replication lag in excess of 15s.

As a result I grabbed the following logs (with comments, less relevant stuff removed, a tiny bit of my logging added):

```
## the backup has already been initiated...
I0102 10:49:47.410406   29387 mysqld.go:408] Mysqld.Shutdown: waiting for socket file (/vitess/mysql/datadir/mysql.sock) and pid file (/vitess/mysql/datadir/mysql.pid) to disappear
I0102 10:49:47.412248   29387 backup.go:413] found 302 files to backup

## ...wating for backup to finish...

I0102 10:50:26.325807   29387 hook.go:118] hook: executing hook: /vitess/vthook/mysqld_start
I0102 10:50:34.520223   29387 backup.go:386] restarting mysql replication

# this is the tablet setting itself from backup to restore (PR addition)
I0102 10:50:35.524483   29387 tabletserver.go:427] setServingType(...)
I0102 10:50:35.524502   29387 tabletserver.go:429] tsv.target.TabletType: BACKUP
I0102 10:50:35.524509   29387 tabletserver.go:431] tabletType: RESTORE
I0102 10:50:35.524516   29387 tabletserver.go:432] tsv.state: NOT_SERVING

## note that no action was taken as a result of setting a non-serving state,
## the tablet was _already_ non-serving
E0102 10:50:35.524527   29387 rpc_backup.go:76] SetServingType to NOT_SERVING took no action: <nil>
I0102 10:50:35.530952   29387 state_change.go:130] Executing post-action state refresh: after backup
I0102 10:50:35.535083   29387 state_change.go:167] Running tablet callback because: after backup

## but the tablet then sets itself back to rdonly + serving
I0102 10:50:35.639009   29387 tabletserver.go:427] setServingType(...)
I0102 10:50:35.639038   29387 tabletserver.go:429] tsv.target.TabletType: RESTORE
I0102 10:50:35.639050   29387 tabletserver.go:431] tabletType: RDONLY
I0102 10:50:35.639059   29387 tabletserver.go:432] tsv.state: NOT_SERVING
I0102 10:50:35.639068   29387 tabletserver.go:312] TabletServer state: NOT_SERVING (Not Serving) -> NOT_SERVING (Transitioning)
I0102 10:50:35.639083   29387 tabletserver.go:312] TabletServer state: NOT_SERVING (Transitioning) -> SERVING

## and now it starts a health check...
I0102 10:50:35.639146   29387 rpc_backup.go:95] kicking off health check

## the Backup function exits...
## ... and health chec8k notices the replication lag a few seconds later
I0102 10:50:39.646097   29387 healthcheck.go:269] Disabling query service because of health-check failure: reported replication lag: 18 higher than unhealthy threshold: 15
```

### Secondary goal: Why doesn't it work

Serving status is an attribute that is primarily controlled by small number of things:

1. Intended behavior based on tablet type, e.g., a REPLICA _should_ serve;
2. Derived behavior based on current health, e.g., this REPLICA is unhealthy and _should not_ serve;
3. A call to `SetServingType` that explicitly carries a new serving state.

Anyway, basically once we force a non-serving state in the naive way the `refreshTablet` code path that runs after all state changes gets executed and has (roughly) the following code:

```golang
switch tabletType {
    case MASTER, REPLICA, RDONLY:
        if !tabletIsServing {
            // set the tablet to serving
        }
}
```

The specific call path we follow to hit that is:

```
rpc_backup.go :: Backup
  -> state_change.go :: refreshTablet
       -> state_change.go :: updateState
            -> state_change.go :: changeCallback
```

There are some knobs to control the behavior during startup based on an existing health check error but initial attempts to co-opt that left me with a tablet stranded in RDONLY + NOT_SERVING. I had another idea while writing this up since `changeCallback` has reference to both previous and next tablets but I suspect this will have the same impact (i.e. none) as attempts to plumb some config through this callpath & fiddle with `disallowQueryService`.

If that doesn't work I'm going to call this for the time being.

### (New, potential) Suggested fix

Aside from the idea above I think a reasonable fix might be to build some understanding into the health checkers / aggregators that when we observe the following states for mysqld: `running -> down -> running` that the replication lag data taken from before the daemon went down can't be used to judge the health of the tablet and introduce an error case to represent that.

That said I believe additional work would still need to be done to update the `refreshTablet` chain as it doesn't check for a healthy tablet before setting tablet to serving=true.

**Comment by sougou** [comment]
This is an excellent analysis. I think the root problem is historical. The replication-based health check feature was organically added on top of the existing flow. So, it doesn't seem to cover all the corner cases.

I think your suggested fix should work: deleting previous data when going non-serving, and then going into serving state only after looking at the replication lag.

It may also be worth taking a step back and thinking about how this should have been wired in the first place, and implementing that approach. This way, any new additions won't break what we just coded.

**Comment by setassociative** [comment]
To close the loop on my last idea: I tried explicitly setting serve state to false in changeCallback if the old tablet type was a Backup and disabling the health check explicitly kicked off in Backup. It, also, did not work.

The timer that kicks off the normal health checks just keeps running and as soon as it does it sees that there is on replication lag (incorrectly) pushes us back into serving. In retrospect that should have been expected based on previous analysis ¯\\\_(ツ)_/¯. My last thought is that because of the timer-run health check fixing this is going to require updating both the health check logic _and_ the `refreshTablet` path. Either alone will be stymied by the other.

To @sougou's point about reassessing how this should have been wired:

I don't have sufficient perspective feel authoritative but the root of this confusion seems to be that we've distributed the responsibility for controlling whether or not a tablet is serving to all of the components that have input into the decision without any one having sufficient context to actually make that decision. Some attention to this was paid given the ability to disable serving via `setServicesDesiredState` but it seems relatively specific / permanent in intent.

If I were to rewrite this today it would probably look like a collection of inputs that can be polled for state and a single function in tablet manager responsible for aggregating everything and making a serve/non-serve decision. This would be run on a timer or could be invoked synchronously in the event that something catastrophic happened and we wanted to force a serving state change _now_.

In any case I'm out of my timebox for this and, I think, the relative priority of a few seconds of potentially old data while the health check catches up is going to push me to something else.

**Comment by deepthi** [comment]
I followed the instructions in #5000 and reproduced this on a local cluster. Here's a selection from debug log of a replica:
1. Replica is lagging by 14s before we start a backup
```I0815 14:17:36.904421    3552 flavor.go:255] 4426: seconds behind master 10
I0815 14:17:36.904444    3552 replication_reporter.go:87] Slave status: {f4bea0f8-bfa1-11e9-8ada-34e12d1e6711:1-11726 true true 10 deepthi-ThinkPad 17100 10}
I0815 14:17:36.904465    3552 replication_reporter.go:105] 4426: replication reporter value 10
I0815 14:17:36.904476    3552 health.go:136] 4426: aggregated delay 10s
I0815 14:17:41.905723    3552 flavor.go:255] 4426: seconds behind master 14
I0815 14:17:41.905770    3552 replication_reporter.go:87] Slave status: {f4bea0f8-bfa1-11e9-8ada-34e12d1e6711:1-14910 true true 14 deepthi-ThinkPad 17100 10}
I0815 14:17:41.905820    3552 replication_reporter.go:105] 4426: replication reporter value 14
I0815 14:17:41.905858    3552 health.go:136] 4426: aggregated delay 14s
I0815 14:17:45.767206    3552 state_change.go:134] Executing post-action state refresh: before backup
I0815 14:17:45.771185    3552 state_change.go:169] Running tablet callback because: before backup
I0815 14:17:45.771202    3552 state_change.go:315] Disabling query service on type change, reason: not a serving tablet type(BACKUP)
I0815 14:17:45.771211    3552 tabletserver.go:360] TabletServer state: SERVING -> SHUTTING_DOWN (Shutting Down)
```
2. Lag is 16s when we stop the slave
```I0815 14:17:45.774764    3552 builtinbackupengine.go:250] getting current replication status
I0815 14:17:45.775117    3552 flavor.go:255] 4426: seconds behind master 16
I0815 14:17:45.776987    3552 hook.go:171] preflight_stop_slave hook doesn't exist
I0815 14:17:45.777170    3552 query.go:73] exec STOP SLAVE
I0815 14:17:45.781811    3552 flavor.go:255] 4426: seconds behind master 
```
3. Once the slave is stopped, `show slave status` doesn't give us a valid value for seconds_behind_master, so the last log line above has a blank field
4. After backup, we exec `START SLAVE` and at this point seconds behind master is 0 (!!!)
```I0815 14:17:50.998925    3552 builtinbackupengine.go:321] restarting mysql replication
I0815 14:17:50.999769    3552 query.go:73] exec START SLAVE
I0815 14:17:51.001509    3552 hook.go:171] postflight_start_slave hook doesn't exist
I0815 14:17:51.003149    3552 flavor.go:255] 4426: seconds behind master 0
```
5. During tablet callback, we are calling replicationReporter.Report, and SBM is still 0
```I0815 14:17:51.015903    3552 state_change.go:169] Running tablet callback because: after backup
I0815 14:17:51.017988    3552 flavor.go:255] 4426: seconds behind master 0
I0815 14:17:51.018019    3552 replication_reporter.go:87] Slave status: {f4bea0f8-bfa1-11e9-8ada-34e12d1e6711:1-17394 true true 0 deepthi-ThinkPad 17100 10}
I0815 14:17:51.018052    3552 replication_reporter.go:105] 4426: replication reporter value 0
I0815 14:17:51.018076    3552 health.go:136] 4426: aggregated delay 0s
I0815 14:17:51.018094    3552 state_change.go:222] 4426: replication delay 0s
```
6. We run a healthcheck before returning from the call to backup, SBM is still 0
```I0815 14:17:51.018795    3552 state_change.go:158] Done with post-action state refresh
I0815 14:17:51.019649    3552 flavor.go:255] 4426: seconds behind master 0
I0815 14:17:51.019668    3552 replication_reporter.go:87] Slave status: {f4bea0f8-bfa1-11e9-8ada-34e12d1e6711:1-17394 true true 0 deepthi-ThinkPad 17100 10}
I0815 14:17:51.019688    3552 replication_reporter.go:105] 4426: replication reporter value 0
I0815 14:17:51.019703    3552 health.go:136] 4426: aggregated delay 0s
I0815 14:17:51.019731    3552 rpc_server.go:95] TabletManager.Backup(concurrency:4 )(on zone1-0000000101 from ): <nil>
```
7. It's only the next healthcheck (5 seconds later) that picks up the lag correctly:
```I0815 14:17:51.020845    3552 flavor.go:255] 4426: seconds behind master 0
I0815 14:17:51.020871    3552 replication_reporter.go:87] Slave status: {f4bea0f8-bfa1-11e9-8ada-34e12d1e6711:1-17394 true true 0 deepthi-ThinkPad 17100 10}
I0815 14:17:51.020900    3552 replication_reporter.go:105] 4426: replication reporter value 0
I0815 14:17:51.020914    3552 health.go:136] 4426: aggregated delay 0s
I0815 14:17:56.021774    3552 flavor.go:255] 4426: seconds behind master 26
I0815 14:17:56.021795    3552 replication_reporter.go:87] Slave status: {f4bea0f8-bfa1-11e9-8ada-34e12d1e6711:1-20691 true true 26 deepthi-ThinkPad 17100 10}
I0815 14:17:56.021817    3552 replication_reporter.go:105] 4426: replication reporter value 26
I0815 14:17:56.021832    3552 health.go:136] 4426: aggregated delay 26s
```
From this, it seems like `SHOW SLAVE STATUS` is itself not returning the right SBM value for a few seconds after `START SLAVE`.

**Comment by deepthi** [comment]
more detailed log from next run:
```I0815 14:36:09.959737   11844 state_change.go:134] Executing post-action state refresh: after backup
I0815 14:36:09.967325   11844 state_change.go:169] Running tablet callback because: after backup
I0815 14:36:09.970162   11844 flavor_mysql.go:108] 4426: show slave status returned: map[Master_Port:17100 Relay_Log_Space:18959961 Maste
r_SSL_Cert: Master_SSL_Verify_Server_Cert:No Master_Info_File:mysql.slave_master_info Master_TLS_Version: Until_Condition:None Master_SSL
_Key: Last_SQL_Error_Timestamp: Replicate_Rewrite_DB: Last_Error: Master_SSL_CA_Path: Last_IO_Error: Master_User:vt_repl Relay_Log_File:v
t-0000000101-relay-bin.000003 Relay_Master_Log_File:vt-0000000100-bin.000001 Replicate_Wild_Ignore_Table: Master_SSL_CA_File: SQL_Remaini
ng_Delay: Master_Retry_Count:86400 Slave_SQL_Running_State:Slave has read all relay log; waiting for more updates Master_SSL_Crl: Master_
Log_File:vt-0000000100-bin.000001 Replicate_Ignore_Server_Ids: Executed_Gtid_Set:7b14a4ab-bfa4-11e9-90c4-34e12d1e6711:1-25591 Auto_Positi
on:1 Master_Host:deepthi-ThinkPad Master_UUID:7b14a4ab-bfa4-11e9-90c4-34e12d1e6711 Retrieved_Gtid_Set: Slave_IO_State:Waiting for master 
to send event Relay_Log_Pos:4 Replicate_Wild_Do_Table: Replicate_Do_Table: Seconds_Behind_Master:0 Last_IO_Error_Timestamp: Connect_Retry
:10 Replicate_Ignore_DB: SQL_Delay:0 Read_Master_Log_Pos:7395960 Skip_Counter:0 Last_SQL_Errno:0 Master_Server_Id:938056168 Replicate_Do_
DB: Master_SSL_Allowed:No Last_IO_Errno:0 Last_Errno:0 Until_Log_File: Until_Log_Pos:0 Master_SSL_Cipher: Master_Bind: Slave_IO_Running:Y
es Exec_Master_Log_Pos:7395960 Last_SQL_Error: Slave_SQL_Running:Yes Replicate_Ignore_Table: Master_SSL_Crlpath: Channel_Name:]
I0815 14:36:09.970219   11844 flavor.go:255] 4426: seconds behind master 0
...
I0815 14:36:09.970821   11844 state_change.go:158] Done with post-action state refresh
I0815 14:36:09.971514   11844 flavor_mysql.go:108] 4426: show slave status returned: map[Master_User:vt_repl Replicate_Ignore_Server_Ids: Master_Bind: Last_IO_Error_Timestamp: Replicate_Wild_Ignore_Table: Master_SSL_Cert: Seconds_Behind_Master:0 Last_IO_Errno:0 Slave_SQL_Running_State:Slave has read all relay log; waiting for more updates Master_SSL_Crlpath: Until_Condition:None Executed_Gtid_Set:7b14a4ab-bfa4-11e9-90c4-34e12d1e6711:1-25591 Replicate_Rewrite_DB: Replicate_Ignore_DB: Last_Errno:0 Retrieved_Gtid_Set: Slave_IO_Running:Yes Master_SSL_Key: Read_Master_Log_Pos:7395960 Last_Error: Exec_Master_Log_Pos:7395960 SQL_Remaining_Delay: Connect_Retry:10 Master_SSL_Allowed:No Replicate_Ignore_Table: SQL_Delay:0 Master_Retry_Count:86400 Relay_Log_File:vt-0000000101-relay-bin.000003 Relay_Master_Log_File:vt-0000000100-bin.000001 Master_SSL_CA_File: Last_SQL_Error: Master_TLS_Version: Replicate_Do_Table: Last_SQL_Errno:0 Last_SQL_Error_Timestamp: Master_Info_File:mysql.slave_master_info Master_Port:17100 Replicate_Do_DB: Replicate_Wild_Do_Table: Until_Log_Pos:0 Master_SSL_CA_Path: Master_SSL_Cipher: Master_UUID:7b14a4ab-bfa4-11e9-90c4-34e12d1e6711 Channel_Name: Master_Server_Id:938056168 Auto_Position:1 Slave_IO_State:Waiting for master to send event Master_Host:deepthi-ThinkPad Relay_Log_Pos:4 Master_SSL_Verify_Server_Cert:No Master_Log_File:vt-0000000100-bin.000001 Slave_SQL_Running:Yes Relay_Log_Space:18959961 Skip_Counter:0 Until_Log_File: Last_IO_Error: Master_SSL_Crl:]
I0815 14:36:09.971555   11844 flavor.go:255] 4426: seconds behind master 0
...
I0815 14:36:09.971617   11844 rpc_server.go:95] TabletManager.Backup(concurrency:4 )(on zone1-0000000101 from ): <nil>
I0815 14:36:09.972398   11844 flavor_mysql.go:108] 4426: show slave status returned: map[Master_Server_Id:938056168 Replicate_Do_Table: Replicate_Wild_Ignore_Table: Last_IO_Errno:0 Master_Info_File:mysql.slave_master_info Last_IO_Error_Timestamp: Auto_Position:1 Last_Error: Until_Log_Pos:0 Master_UUID:7b14a4ab-bfa4-11e9-90c4-34e12d1e6711 Seconds_Behind_Master:0 Replicate_Ignore_Table: Last_Errno:0 Until_Condition:None Master_Retry_Count:86400 Skip_Counter:0 Exec_Master_Log_Pos:7395960 Until_Log_File: Master_Host:deepthi-ThinkPad Replicate_Do_DB: Master_SSL_Crlpath: Slave_SQL_Running:Yes Last_SQL_Errno:0 Executed_Gtid_Set:7b14a4ab-bfa4-11e9-90c4-34e12d1e6711:1-25591 Master_SSL_Crl: Retrieved_Gtid_Set: Slave_IO_Running:Yes Master_SSL_CA_Path: SQL_Delay:0 Connect_Retry:10 Master_Log_File:vt-0000000100-bin.000001 Last_SQL_Error_Timestamp: Replicate_Rewrite_DB: Master_SSL_Cert: Master_SSL_Cipher: Relay_Log_Pos:4 Relay_Master_Log_File:vt-0000000100-bin.000001 Relay_Log_Space:18959961 Master_SSL_CA_File: Last_IO_Error: Replicate_Ignore_Server_Ids: SQL_Remaining_Delay: Slave_SQL_Running_State:Slave has read all relay log; waiting for more updates Master_Port:17100 Replicate_Wild_Do_Table: Master_Bind: Slave_IO_State:Waiting for master to send event Last_SQL_Error: Channel_Name: Master_SSL_Allowed:No Master_SSL_Key: Replicate_Ignore_DB: Master_TLS_Version: Master_User:vt_repl Relay_Log_File:vt-0000000101-relay-bin.000003 Read_Master_Log_Pos:7395960 Master_SSL_Verify_Server_Cert:No]
I0815 14:36:09.972444   11844 flavor.go:255] 4426: seconds behind master 0
I0815 14:36:09.972454   11844 replication_reporter.go:87] Slave status: {7b14a4ab-bfa4-11e9-90c4-34e12d1e6711:1-25591 true true 0 deepthi-ThinkPad 17100 10}
I0815 14:36:09.972468   11844 replication_reporter.go:105] 4426: replication reporter value 0
I0815 14:36:09.972476   11844 health.go:136] 4426: aggregated delay 0s
I0815 14:36:14.981131   11844 flavor_mysql.go:108] 4426: show slave status returned: map[Master_Host:deepthi-ThinkPad Relay_Master_Log_File:vt-0000000100-bin.000001 Replicate_Wild_Do_Table: Skip_Counter:0 Until_Log_Pos:0 Master_SSL_CA_Path: Master_SSL_Verify_Server_Cert:No SQL_Remaining_Delay: Master_SSL_Allowed:No Last_IO_Error: Last_SQL_Errno:0 Master_Info_File:mysql.slave_master_info Relay_Log_Pos:198114 Until_Log_File: Master_SSL_Crlpath: Replicate_Rewrite_DB: Until_Condition:None Master_Port:17100 Last_SQL_Error: Replicate_Ignore_Server_Ids: Master_TLS_Version: Slave_IO_State:Waiting for master to send event Master_SSL_Cert: Master_Server_Id:938056168 Master_SSL_Crl: Channel_Name: Last_Errno:0 Connect_Retry:10 Slave_IO_Running:Yes Replicate_Ignore_Table: Exec_Master_Log_Pos:7593636 Seconds_Behind_Master:38 Master_UUID:7b14a4ab-bfa4-11e9-90c4-34e12d1e6711 SQL_Delay:0 Read_Master_Log_Pos:19473559 Master_SSL_Cipher: Replicate_Do_DB: Last_Error: Last_IO_Errno:0 Slave_SQL_Running_State:System lock Auto_Position:1 Relay_Log_Space:12078252 Master_Bind: Last_IO_Error_Timestamp: Retrieved_Gtid_Set:7b14a4ab-bfa4-11e9-90c4-34e12d1e6711:25592-67382 Master_User:vt_repl Relay_Log_File:vt-0000000101-relay-bin.000004 Last_SQL_Error_Timestamp: Executed_Gtid_Set:7b14a4ab-bfa4-11e9-90c4-34e12d1e6711:1-26275 Master_Log_File:vt-0000000100-bin.000001 Replicate_Ignore_DB: Master_SSL_Key: Master_Retry_Count:86400 Slave_SQL_Running:Yes Replicate_Do_Table: Master_SSL_CA_File: Replicate_Wild_Ignore_Table:]
I0815 14:36:14.981247   11844 flavor.go:255] 4426: seconds behind master 38
```

**Comment by deepthi** [comment]
After analyzing the detailed log, it seems that SBM starts as 0 after `start slave` and does not get updated to a "correct" value until at least 1 transaction has been applied on the replica.
